sphinx/ext/autodoc/__init__.py

```diff
@@ -1062,56 +1062,34 @@ def document_members(self, all_members: bool = False) -> None:
 
     def add_directive_header(self, sig: str) -> None:
         sourcename = self.get_sourcename()
+        if inspect.is_singledispatch_function(self.object):
+            # Set the flag so add_line will add the signatures
+            # of the other implementations
+            self._singledispatch_flag = True
+
         super().add_directive_header(sig)
 
         if inspect.iscoroutinefunction(self.object):
             self.add_line('   :async:', sourcename)
 
-
-class SingledispatchFunctionDocumenter(FunctionDocumenter):
```

Comment:
Could you keep "singledispatch"ed classes as is? That is needed to keep compatibilities.

Reply:
Please could you expand on this comment. What compatibilities do I need to keep, is it just that SingledispatchFunctionDocumenter must exist, or that I can't change the code in it?

Reply:
Yes, SingledispatchFunctionDocumenter must keep existing.

From the experience so far, there is some possibility that a 3rd party extension accesses such old classes. So we keep them during two major versions and let developers know the deprecation of the components.
https://github.com/sphinx-doc/sphinx/blob/3.x/CONTRIBUTING.rst#id20

Reply:
The registry contains only one documenter for each function type. If we keep that, then autofunction will always point at the FunctionDocumenter. To make it support singledispatch functions, FunctionDocumenter must know how to document them, which means that SingledispatchFunctionDocumenter is effectively an alias for it. I have put it back, but moved the functionality into the baseclass so both can now document singledispatch functions

Reply:
So is:
```python
class SingledispatchFunctionDocumenter(FunctionDocumenter):
    pass
```
enough to maintain backwards compatibility? Even if it is never actually instantiated by sphinx?

Reply:
Yes. It is what I needed. After merging this, I'll add deprecation warning to them. And they will be removed in future.

Reply:
Could you remove `    app.add_autodocumenter(SingledispatchFunctionDocumenter)` entries from setup function?